### PR TITLE
【SCB-1615】Extended dynamic configuration support for Nacos

### PR DESCRIPTION
**Pull request by 160731liupf**
How to use:
1.Download base model from http://start.servicecomb.io and add maven dependence:

org.apache.servicecomb
config-nacos
2.0.0-SNAPSHOT

2.Start nacos console model(git url:https://github.com/alibaba/nacos) ,both account and password is "nacos" and add properties(example):
Data ID: example
Group:DEFAULT_GROUP
JSON:
{
"nacos":"666"
}
3.In base model,add info in microservice.yaml:
nacos:
  config:
    serverAddr: 127.0.0.1:8848
    dataId: example
    group: DEFAULT_GROUP
4.Then add blow code and start base model,you will get properties(If properties on nacos has changed, you can also get new value):
@RestSchema(schemaId = "nacos")
@RequestMapping(path = "/")
public class NacosImpl {
/**
* 从nacos获取配置
* @return
*/
@GetMapping(path = "/config")
@Responsebody
public String config() throws Exception{
final String config = DynamicPropertyFactory.getInstance()
.getStringProperty("nacos","").getValue();
return JSON.toJSONString(config);
}
}

### CODE REVIEW

File: dynamic-config/config-apollo/pom.xml [hunk 1]
```diff
@@ -44,5 +44,9 @@
       <groupId>javax.validation</groupId>
       <artifactId>validation-api</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-jdk14</artifactId>
+    </dependency>
```
Comment (liubao68) [inline dynamic-config/config-apollo/pom.xml]: Why you modifiy this? I think can be deleted.

File: dynamic-config/config-cc/pom.xml [hunk 2]
```diff
@@ -51,6 +51,11 @@
       <artifactId>vertx-codegen</artifactId>
       <scope>provided</scope>
     </dependency>
+
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-jdk14</artifactId>
+    </dependency>
```
Comment (liubao68) [inline dynamic-config/config-cc/pom.xml]: same as above

File: dynamic-config/config-nacos/pom.xml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    </dependency>
+    <!--nacos client-->
+    <dependency>
+      <groupId>com.alibaba.nacos</groupId>
+      <artifactId>nacos-client</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-jdk14</artifactId>
+    </dependency>
```
Comment (liubao68) [inline dynamic-config/config-nacos/pom.xml]: Does nacos must use jdk logger? ServiceComb use sl4j to log messages, and users can  choose from jdk logger, log4j, logback, log4j2, etc. It's better not put this dependency

**Comment by liubao68** [comment]
suggestions: you can add a README.md file for this module, and put your usage guide in this file.

File: dynamic-config/config-apollo/pom.xml (hunk 1 above)
Comment (160731liupf) [reply dynamic-config/config-apollo/pom.xml]: Sorry,It's my fault.Mybe my project is not full compilation .A few days ago,When I run test in config-apollo,I found it print nothing.There is no need to add this dependency,I had remove it in a new commit.

File: dynamic-config/config-cc/pom.xml (hunk 2 above)
Comment (160731liupf) [reply dynamic-config/config-cc/pom.xml]: It has removed in a new commit.

File: dynamic-config/config-nacos/pom.xml (hunk 3 above)
Comment (160731liupf) [reply dynamic-config/config-nacos/pom.xml]: I had removed it in a new commit

**Comment by 160731liupf** [comment]
> suggestions: you can add a README.md file for this module, and put your usage guide in this file.

Thanks for your suggestions,I had create a new commit to remove dependency and add README.md file.

File: java-chassis-dependencies/default/pom.xml [hunk 4]
```diff
@@ -1341,6 +1343,11 @@
         <artifactId>business-service-springmvc-archetype</artifactId>
         <version>${project.version}</version>
       </dependency>
+      <dependency>
```
Comment (liubao68) [inline java-chassis-dependencies/default/pom.xml]: Add new libraries need update LICENCE file in ${HOME}/java-chassis-distributioni/release/src/LICENCE

File: dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/client/NacosClient.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+              }
+          }
+          @Override
+          public Executor getExecutor() {
+            return null;
+          }
+        });
+      }catch (Exception e){
+          LOGGER.error("Receive nacos config error: ", e);
+      }
```
Comment (liubao68) [inline dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/client/NacosClient.java]: A small piece of code is not well formatted using code template in ${HOME}/etc

File: dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/client/ConfigurationAction.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.servicecomb.config.client;
+
+public enum ConfigurationAction {
```
Comment (liubao68) [inline dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/client/ConfigurationAction.java]: You used same package and class name with config-appolo. I Suggest you refactor all code to use a different package name for all class. e.g. 
```
org.apache.servicecomb.config.nacos
```
And we will handle other implentations, maybe in future.

File: dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/archaius/sources/NacosConfigurationSourceImpl.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+import com.netflix.config.WatchedUpdateResult;
+
+public class NacosConfigurationSourceImpl implements ConfigCenterConfigurationSource {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ConfigCenterConfigurationSource.class);
+
+  private final Map<String, Object> valueCache = new ConcurrentHashMap<>();
+
+  private List<WatchedUpdateListener> listeners = new CopyOnWriteArrayList<>();
+
+  private static final String NACOS_CONFIG_SERVER_ADDR_KEY = "nacos.config.serverAddr";
```
Comment (liubao68) [inline dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/archaius/sources/NacosConfigurationSourceImpl.java]: Can use the same constant in NacosConfig

File: java-chassis-dependencies/default/pom.xml (hunk 4 above)
Comment (160731liupf) [reply java-chassis-dependencies/default/pom.xml]: I had add 
"* nacos-client (com.alibaba.nacos:1.1.4 - https://github.com/alibaba/nacos)"
 in LICENSE file.

File: dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/client/ConfigurationAction.java (hunk 6 above)
Comment (160731liupf) [reply dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/client/ConfigurationAction.java]: The package name of config-nacos model has replaced with "org.apache.servicecomb.config.nacos".

File: dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/archaius/sources/NacosConfigurationSourceImpl.java (hunk 7 above)
Comment (160731liupf) [reply dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/archaius/sources/NacosConfigurationSourceImpl.java]: The name of key is replaced with "SERVER_ADDR".

File: dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/client/NacosClient.java (hunk 5 above)
Comment (160731liupf) [reply dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/client/NacosClient.java]: 可以用中文解释一下吗？我不明白要怎么修改比较好

Comment (liubao68) [reply dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/client/NacosClient.java]: 有少量代码没有正确格式化（比如空格等），可以使用IDE自动格式化一下。 格式化模板在servicecomb的 etc 目录。 里面包含了eclipse, IDEA等IDE的格式化模板。

Comment (160731liupf) [reply dynamic-config/config-nacos/src/main/java/org/apache/servicecomb/config/client/NacosClient.java]: 你好，config-nacos 模块我已经用intellij-java-google-style.xml格式化好了，这是pr地址：https://github.com/apache/servicecomb-java-chassis/pull/1414/files

**Pull request merged**